### PR TITLE
POST encoding to 'application/x-www-form-urlencoded' per OAuth 2.0 spec

### PR DESCRIPTION
**Pull request by anthony-vito**
The purpose of these changes is compatibility with FusionAuth.

These changes add a configuration key of 'tokenPostEncoding' as a pseudo
enumeration that is either 'multi-part' ( the default ) or 'url-encoded'.
Setting this configuration key to a value of 'url-encoded' will cause
the token requests to be sent as 'application/x-www-form-urlencoded'
POSTs. ( FusionAuth does not support multi-part POSTS for tokens )

This is accomplished by explicitly typing the token request data and
passing that into the function performing the fetch. The actual POST
body is then built along with the proper headers in the function
doing the fetching.

This commit also increases the code verifier random length to 96.
( FusionAuth seems to reject shorter code verifiers )

**EDIT** It looks like the request for a token and refresh token _MUST_ be in 'application/x-www-form-urlencoded' format according to the OAuth 2.0 spec. https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.3. I will update this pull request to remove the configuration option and always use the url-encoded style.

closes issue #14

### CODE REVIEW

**Comment by soofstad** [comment]
Hi, thanks for pointing this out. Was not aware that this was causing any issues.
Appreciate the PR. I will have a closer look as soon as I get the time. Won't be long.
Have you tested this with any other auth provider (Keycloak, Azure, Github etc.)?
Ether way, the goal for the package is to be provider agnostic, and following the RFC's specifications, so this should be included, and optionally disabled if that is needed for some provider.

**Comment by anthony-vito** [comment]
I have not tested the change with other providers. That said it shouldn't break any compatibility. Other providers were likely just "accepting" the multi-part encoding style to be flexible. FusionAuth is just more strict ( probably because it's a pain to handle multi-parts in Java )

In the first attempt I made it a configuration to switch between encodings.... After realizing the url-encoding is required by spec I thought it'd be cleaner to just change it and not ask, since that will be the universally compatible encoding.

File: src/AuthContext.tsx
Comment (soofstad) [inline src/AuthContext.tsx:12]:
```suggestion
import { IAuthContext, IAuthProvider, TInternalConfig, TTokenData, TTokenResponse } from './Types'
```
This was removed with your latest commit, and now caused an import error

File: src/AuthContext.tsx
Comment (soofstad) [inline src/AuthContext.tsx:44]:
```suggestion
```
Same with this. Since you decided to remove it, which I support.

File: src/AuthContext.tsx
Comment (soofstad) [inline src/AuthContext.tsx:54]:
```suggestion
```

File: src/AuthContext.tsx
Comment (soofstad) [inline src/AuthContext.tsx:51]: Good catch :) 
This could potentially create some problems

File: src/authentication.ts
Comment (soofstad) [inline src/authentication.ts:43]: could rename this to "postWithXForm" now

File: src/authentication.ts
Comment (soofstad) [inline src/authentication.ts:41]: This probably works just fine. But I like axios way of url encoding x-forms using URLSearchParams better.
I think it's more readable, and probably handles some edge cases better.
See https://axios-http.com/docs/urlencoded for reference.

File: src/authentication.ts
Comment (soofstad) [inline src/authentication.ts:84]:
```suggestion
  const tokenRequest: TTokenRequestWithCodeAndVerifier = {
```
Since we now have these explicit types, I think it makes sense to use them here.

File: src/authentication.ts
Comment (soofstad) [inline src/authentication.ts:100]:
```suggestion
  const tokenRequest: TTokenRequestForRefresh = {
```
Same here. Should use the explicit type

**Review by soofstad — CHANGES_REQUESTED** [review]
Looks good! Hugely appreciate your contribution.

Tested with other auth providers and they all work fine.

Some minor fixups and a suggestion to not handle the url encoding ourself.
The project is setup with pre-commit. When you think the branch is ready to be merged, you should check if it passes the pre-commit checks as well :slightly_smiling_face:

File: src/authentication.ts
Comment (anthony-vito) [reply src/authentication.ts:41]: I do like those methods, and I have used them. However in this case we have a very restricted set of parameters that will ever be encountered due to the nature of the spec. I prefer this simple method since it doesn't involve any additional libraries!

If it ever fails I'll help re-work it... deal?

File: src/authentication.ts
Comment (anthony-vito) [reply src/authentication.ts]: Yes, agree.

**Comment by anthony-vito** [comment]
I put a few commits on the branch implementing your suggestions. I didn't even realize I left that dead code in there. Need more coffee apparently. Thanks!

File: src/authentication.ts
Comment (soofstad) [reply src/authentication.ts:49]: `URLSearchParams` is a builtin class. No additional library required.
I would still like to see this change, but accept your argument.
I'll leave it up to you :)

**Review by soofstad — APPROVED** [review]

**Comment by soofstad** [comment]
Thanks. Approved the PR. Also took the liberty to cleanup the commit messages before squashing and merging into main.
Will do one last round of testing before creating a new minor release :)

**Pull request merged**
